examples/src/bin/widget_subclass.rs

```diff
@@ -0,0 +1,104 @@
+use std::cell::RefCell;
+use std::env;
+
+use gio::prelude::*;
+use glib::{glib_object_subclass, glib_wrapper, WeakRef};
+use gtk::prelude::*;
+use gtk::subclass::prelude::*;
+
+#[derive(Debug)]
+pub struct ButtonPrivate {
+    child: RefCell<WeakRef<gtk::Widget>>,
+}
+
+impl Drop for ButtonPrivate {
+    fn drop(&mut self) {
+        // Child widgets need to be manually unparented.
+        if let Some(child) = self.child.get_mut().upgrade() {
+            child.unparent();
+        }
+    }
+}
+
+impl ObjectSubclass for ButtonPrivate {
+    const NAME: &'static str = "MyButton";
+    type Type = Button;
+    type ParentType = gtk::Widget;
+    type Instance = glib::subclass::simple::InstanceStruct<Self>;
+    type Class = glib::subclass::simple::ClassStruct<Self>;
+
+    glib_object_subclass!();
+
+    fn class_init(klass: &mut Self::Class) {
+        // The layout manager determines how child widgets are laid out.
+        klass.set_layout_manager_type::<gtk::BinLayout>();
+
+        // Make it look like a GTK button.
+        klass.set_css_name("button");
+    }
+
+    fn new() -> Self {
+        Self {
+            child: RefCell::new(WeakRef::new()),
+        }
+    }
+}
+
+impl ObjectImpl for ButtonPrivate {
+    fn constructed(&self, obj: &Self::Type) {
+        self.parent_constructed(obj);
+
+        // Create the child label.
+        let child = gtk::Label::new(Some("Hello world!"));
+        child.set_parent(obj);
+        *self.child.borrow_mut() = child.upcast::<gtk::Widget>().downgrade();
+
+        // Make it look like a GTK button with a label (as opposed to an icon).
+        obj.add_css_class("text-button");
+
+        // Connect a gesture to handle clicks.
+        let gesture = gtk::GestureClick::new();
+        gesture.connect_released(|gesture, _, _, _| {
+            gesture.set_state(gtk::EventSequenceState::Claimed);
+            println!("Button pressed!");
+        });
+        obj.add_controller(&gesture);
+    }
+}
+
+impl WidgetImpl for ButtonPrivate {}
```

Comment:
We tend to put the private struct in a imp module

Reply:
Basically a logical extension of https://doc.rust-lang.org/1.0.0/style/style/naming/README.html#avoid-redundant-prefixes-[rfc-356] . Use module namespaces instead of putting namespaces into type names.

Reply:
Updated.